### PR TITLE
doc: describe compile-time options

### LINKED ISSUE #8572: Document compile-time configuration parameters
There are a few things that can be set at compile-time and change the behavior of the libraries:

- GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
- GOOGLE_CLOUD_CPP_LOGGING_MIN_SEVERITY_ENABLED
- GOOGLE_CLOUD_CPP_HAVE_COROUTINES (soon)

Probably the `packaging.md` document or something like it.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #8572

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10822)

### CODE REVIEW

File: doc/compile-time-configuration.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
+# Compile time configuration for \`google-cloud-cpp
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:1]: closing backtick?

File: doc/compile-time-configuration.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+Some features represent groups of libraries. For example, `__ga_libraries__`
+requests all the GA libraries, and `__experimental_libraries__` represent the
+libraries that are not yet GA.
+
+A few features enable experimental functionality. We do not expect that
+customers will need to use these. If you have specific questions please start
+a [GitHub Discussion]. With that said:
+
+- `internal-docfx` enables an internal-only tool to generate documentation.
+- `generator` enables an internal-only too to generate new libraries.
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:38]: s/too/tool/

File: doc/compile-time-configuration.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+- `experimental-storage-grpc` enables the GCS+gRPC plugin. Contact your account
+  team if you want to use this feature or are interested in the GA timeline.
+- `experimental-http-transcoding` enables support for HTTP/1.1 transport (as
+  opposed to gRPC over HTTP/2) in some libraries.
+- `experimental-opentelemetry` enables support for [Open Telemetry].
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:43]: s/Open Telemetry/OpenTelemetry/

File: doc/compile-time-configuration.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+## Bazel Options
+
+We expect that application developers using Bazel will include
+`google-cloud-cpp` as an external dependency.
+
+### The C++ Standard
+
+The default Bazel toolchain forces C++11 on Linux and macOS. `google-cloud-cpp`
+requires C++14, you will need to update the C++ standard version. You must
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:74]: s/, you/, so you/

File: doc/compile-time-configuration.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+either:
+
+- Provide your own C++ toolchain configuration for Bazel. Consult the Bazel
+  documentation if you want to attempt this approach
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:78]: period at end of line

File: doc/compile-time-configuration.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+- Add `--cxxopt=-std=c++14` and `--host_cxxopt=-std=c++14` (or higher) in your
+  Bazel command-line
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:80]: period at end of line

File: doc/compile-time-configuration.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+- Add the same options in your Bazel `.bazelrc` file.
+
+The `--host_cxxopt` may be unfamiliar. This is required to support Protobuf
+and gRPC, where the
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:84]: s/, where the/, which compile a "host" version of \`protoc\`./

File: doc/compile-time-configuration.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+### Disabling OpenTelemetry
+
+[Open Telemetry] is enabled by default.  Turning this off may reduce your build
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:88]: s/Open Telemetry/OpenTelemetry/

File: doc/compile-time-configuration.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+times but will also lose the benefits of instrumenting the libraries for
+distributed tracing.  Add `--//:experimental-opentelementry=false` to your
+Bazel command-line parameters to disable Open Telemetry.
+
+[ccmake]: https://cmake.org/cmake/help/latest/manual/ccmake.1.html
+[github discussion]: https://github.com/googleapis/google-cloud-cpp/discussions
+[open telemetry]: https://opentelemetry.io/
```
Comment (dbolduc) [inline doc/compile-time-configuration.md:95]: s/open telemetry/opentelemetry/

**Review by dbolduc — APPROVED** [review]

File: doc/compile-time-configuration.md (hunk 5 above)
Comment (coryan) [reply doc/compile-time-configuration.md]: Done.

File: doc/compile-time-configuration.md (hunk 2 above)
Comment (coryan) [reply doc/compile-time-configuration.md]: Fixed.

File: doc/compile-time-configuration.md (hunk 1 above)
Comment (coryan) [reply doc/compile-time-configuration.md]: Removed

**Pull request merged**